lib/charms/prometheus_k8s/v1/prometheus.py

```diff
@@ -317,99 +318,72 @@ def __init__(self, charm, name, consumes, multi=False):
-                provider. This is only necessary if the
-                `PrometheusConsumer` has been instantiated in
-                `multi` mode.
+        When a consumer charm is related to a Prometheus provider, the
+        consumer sets metadata related to its own scrape
+        configutation.  This metadata is set using Juju application
+        data.  In addition each of the consumer units also sets its own
+        host address in Juju unit relation data.
         """
-        if rel_id is None:
-            rel_id = self.relation_id
+        event.relation.data[self._charm.unit]["prometheus_scrape_host"] = str(
+            self._charm.model.get_binding(event.relation).network.bind_address
+        )
 
-        targets = self._stored.targets.get(rel_id, [])
-        if address in targets:
+        if not self._charm.unit.is_leader():
             return
 
-        target = address + ":" + str(port)
-        targets.append(target)
-        self._update_targets(targets, rel_id)
-
-    def remove_endpoint(self, address, port=80, rel_id=None):
-        """Remove an endpoint from the list of Prometheus scrape targets.
-        Args:
-            address: a string host address (usually IP) of the endpoint that
-                that must be excluded from being monitored by Prometheus.
-            port: an optional (default 80) integer providing the port
-                on which the scrapped endpoint exposes its Prometheus
-                metrics.
-            rel_id: an optional integer providing the relation ID for
-                the related Prometheus monitoring service
-                provider. This is only necessary if the
-                `PrometheusConsumer` has been instantiated in
-                `multi` mode.
-        """
-        if rel_id is None:
-            rel_id = self.relation_id
-
-        targets = self._stored.targets.get(rel_id, [])
-        target = address + ":" + str(port)
-        if target not in targets:
-            return
+        event.relation.data[self._charm.app]["prometheus_scrape_metadata"] = json.dumps(
+            self._scrape_metadata
+        )
 
-        targets.remove(target)
-        self._update_targets(targets, rel_id)
+    def _set_unit_ip(self, event):
+        """Set unit host address
 
-    @property
-    def endpoints(self, rel_id=None):
-        """Returns a list of Prometheus scrape target endpoints.
-        Args:
-            rel_id: an optional integer providing the relation ID for
-                the related Prometheus monitoring service
-                provider. This is only necessary if the
-                `PrometheusConsumer` has been instantiated in
-                `multi` mode.
+        Each time a consumer charm container is restarted it updates its own
+        host address in the unit relation data for the Prometheus provider.
         """
+        for relation in self._charm.model.relations[self._relation_name]:
```

Comment:
is it guaranteed in this context that `relations[self._relation_name]` never raises a KeyError?

Reply:
Yes. See [here](https://github.com/canonical/operator/blob/92c7922e451b14184ceb9e7170d1564306fa1046/ops/model.py#L414). `meta.relations` (which ultimately goes to `RelationMapping` as `relation_meta` and is iterated over to set the keys) comes from `metadata.yaml`.